### PR TITLE
Initial support of writable_search from quickfix

### PR DESCRIPTION
**Pull request by p0deje**
Related to #6.

So, at the moment it just allows to `call writable_search#Parse()` from quickfix window and it will properly create new buffer with writable_search.

I'm not really happy with the code (it seems hacky to me). I just wanted to show the progress and hear your thoughts on this.

P.S. Includes some slight refactoring of parser which makes diff noisy :)

### CODE REVIEW

File: autoload/writable_search/parser.vim [hunk 1]
```diff
@@ -101,6 +111,23 @@ function! s:BuildProxies(grouped_lines)
   return proxies
 endfunction
 
+function! s:BuildProxiesFromQuickfix()
+  let proxies = []
+  let quickfix_list = getqflist()
+
+  for location in quickfix_list
+    let current_proxy = writable_search#proxy#New(location.bufnr)
```
Comment (AndrewRadev) [inline autoload/writable_search/parser.vim]: Hmm, it seems like this shouldn't work. If you're creating a new proxy in the loop, doesn't it end up with only one entry?

Comment (AndrewRadev) [inline autoload/writable_search/parser.vim]: Ah, no, wait, my bad, you're just using one per patch. Completely forgot the implementation of my own plugin :D.

**Comment by AndrewRadev** [comment]
I think it looks pretty good so far. The parsing seems to work, although updating the lines doesn't change the original. But I guess you're probably working on that next :). I also very much appreciate the tests, keep up the good work :).

File: autoload/writable_search/parser.vim (hunk 1 above)
Comment (p0deje) [inline autoload/writable_search/parser.vim]: Had this problem while implementing, but hail to the tests :)

**Comment by p0deje** [comment]
@AndrewRadev Thanks, I'll get back with updates soon.

**Comment by p0deje** [comment]
I'm a little bit confused about how to make file save working. It seems to me that I have to define `b:command` to make ftplugin work correctly, but it's hacky to me.

Do you have any suggestions on how to implement it or refactor existing code?

**Comment by AndrewRadev** [comment]
Hmm, I definitely think some restructuring is in order.

One problem I run into immediately is that `parent_buffer` is not the buffer of the original file, but the writable search buffer. I should think of either renaming it or documenting it a bit. Probable the former, "parent_buffer" doesn't really seem right now that I think about it. Maybe I need a word for that. "Common buffer"? "Result buffer"? I'm open to suggestions.

However, fixing this means taking the new writable search buffer's id. But, in the quickfix example, running `writable_search#Parse()` needs to *first* run the parsing code, and *then* open a new buffer. So it's a bit of a chicken and egg situation.

In the "normal" path, I think this is the sequence of events:

- `:WritableSearch` -- the main entry point
- `writable_search#Start(query, count)` -- this takes a query and prepares it for the command (or uses a default)
- `s:NewBuffer()` -- opens a new empty buffer (only if it's not already in a result buffer)
- `s:Grep(query)` -- runs the query
- `writable_search#Parse()` -- parses the result buffer and renders the new, parsed result. Basically a wrapper for:
- `let b:proxies = writable_search#parser#Run()`, followed by
- `writable_search#Render()`

For the quickfix case, you've modified the `writable_search#parser#Run()` function (or its helper functions, actually), and this looked fine to me to begin with, but now I realize it's probably a bad idea. The parser expects to be in a buffer with the grep results already there.

So, I think we need a new, parallel branch of code that runs the quickfix stuff. First of all, we don't need `#Start()`, since that takes a search query, and we don't need that. So maybe it could work like this:

- `:WritableSearchFromQuickfix` -- the main entry point (maybe with a different name, or maybe `:WritableSearch` in the quickfix buffer could trigger this?)
- `s:NewBuffer()`
- `writable_search#ParseQuickfix()`, which calls:
- `let b:proxies = writable_search#quickfix#Parse()`, followed by
- `writable_search#Render()`

I don't like the slightly inconsistent naming -- `writable_search#parser#Run` vs `writable_search#quickfix#Parse`, so maybe `writable_search#quickfix_parser#Run`? Not sure, I'd appreciate your thoughts on this.

In any case, I think if you move the contents of the helper method to their own file, and you make a new function `#ParseQuickfix`, which is almost identical to `#Parse` (maybe with different names), that could work out.

What do you think?

**Comment by p0deje** [comment]
I've already said that the code is pretty hacky, so the restructuring you propose will eliminate that. I'll try to do that and will let you know once it's ready.

Thanks for your comment!

**Comment by AndrewRadev** [comment]
Hello again :). I've managed to add quickfix support myself at some point (finally got around to it!). Because of this, I guess I'll close the PR for the moment. Thank you very much for your work regardless, it pushed me to put in the effort, if nothing else. And I definitely like the functionality myself :).

**Comment by p0deje** [comment]
Thank you for adding this feature, it's really useful!

**Pull request closed without merging**